File: python/tests/test_neuron.py
"""test_neuron.py — the Python Neuron port.

Mirrors the C# CircleAI.Tests Neuron suite: the concierge decision table + gate,
the two-slot admission gate + eviction, the router-gated slot selection inside
AIService (specialist hot-load, generalist floor), and the NeuronNode facade.
"""
from __future__ import annotations

import os

from circle_ai.device.device_probe import DeviceTier
from circle_ai.hosting import (
    AIOptions,
    AIService,
    ChatTurn,
    HeuristicNeuronRouter,
    NeuronGate,
    NeuronNode,
    NullChatRuntime,
    Organ,
    ResidentSlotManager,
    RouteContext,
    RouteDecision,
    SlotOutcome,
)
from circle_ai.inference.inference import ChatCapability, ModelSelection


# ── test doubles ─────────────────────────────────────────────────────────────


class _FixedGen:
    """Minimal IChatGenerator that returns a fixed reply and records disposal."""

    def __init__(self, reply: str) -> None:
        self._reply = reply
        self.disposed = False

    async def generate_async(self, messages, options=None) -> str:
        return self._reply

    async def stream_async(self, messages, options=None):
        yield self._reply

    async def save_session_async(self, path: str) -> bool:
        with open(path, "w", encoding="utf-8") as fh:
            fh.write("circleai-session-marker\n")
        return True

    async def load_session_async(self, path: str) -> bool:
        if not os.path.isfile(path):
            return False
        with open(path, "r", encoding="utf-8") as fh:
            return fh.read().startswith("circleai-session-marker")

    def dispose(self) -> None:
        self.disposed = True


class _FixedRouter:
    def __init__(self, decision: RouteDecision) -> None:
        self._decision = decision

    def route(self, context: RouteContext) -> RouteDecision:
        return self._decision


class _FixedSelector:
    def __init__(self, selection: ModelSelection) -> None:
        self._selection = selection
        self.calls = 0

    def best_fit(self, probe, required: ChatCapability = ChatCapability.DEFAULT) -> ModelSelection:
        self.calls += 1
        return self._selection

    def all_candidates(self, probe):
        return [self._selection]


def _sel(model_id: str, estimated_bytes: int) -> ModelSelection:
    return ModelSelection(model_id, False, estimated_bytes, DeviceTier.DESKTOP)


# ── concierge router + gate ──────────────────────────────────────────────────


def test_plain_query_routes_to_generalist() -> None:
    d = HeuristicNeuronRouter().route(RouteContext("what's the weather today?"))
    assert d.organ == Organ.GENERALIST
    assert d.capability == ChatCapability.DEFAULT


def test_image_routes_to_vision_specialist() -> None:
    d = HeuristicNeuronRouter().route(RouteContext("what is this?", has_image=True))
    assert d.organ == Organ.SPECIALIST
    assert d.capability == ChatCapability.VISION


def test_reasoning_cue_routes_to_reasoning_specialist() -> None:
    d = HeuristicNeuronRouter().route(RouteContext("please debug this stack trace"))
    assert d.organ == Organ.SPECIALIST
    assert d.capability == ChatCapability.REASONING


def test_long_prompt_routes_to_long_context_specialist() -> None:
    d = HeuristicNeuronRouter(long_context_chars=50).route(RouteContext("x" * 60))
    assert d.organ == Organ.SPECIALIST
    assert d.capability == ChatCapability.LONG_CONTEXT


def test_gate_vetoes_specialist() -> None:
    router = HeuristicNeuronRouter(NeuronGate(allow_specialist=lambda q: False))
    d = router.route(RouteContext("solve this equation"))
    assert d.organ == Organ.GENERALIST


# ── resident slot manager ────────────────────────────────────────────────────


async def test_slot_admits_within_budget() -> None:
    mgr = ResidentSlotManager(1_000, lambda: 1_000_000)
    g = _FixedGen("S")
    a = await mgr.ensure_specialist_async(_sel("spec", 5_000), lambda mid: g)
    assert a.outcome == SlotOutcome.ADMITTED
    assert a.generator is g
    assert mgr.resident_specialist_model_id == "spec"


async def test_slot_denies_over_budget() -> None:
    mgr = ResidentSlotManager(900_000, lambda: 1_000_000)
    a = await mgr.ensure_specialist_async(_sel("spec", 500_000), lambda mid: _FixedGen("S"))
    assert a.outcome == SlotOutcome.INSUFFICIENT_RAM
    assert a.generator is None
    assert mgr.resident_specialist_model_id is None


async def test_slot_already_resident_does_not_rebuild() -> None:
    mgr = ResidentSlotManager(0, lambda: 1_000_000)
    builds: list[str] = []

    def build(mid: str) -> _FixedGen:
        builds.append(mid)
        return _FixedGen("S")

    await mgr.ensure_specialist_async(_sel("spec", 1), build)
    second = await mgr.ensure_specialist_async(_sel("spec", 1), build)
    assert second.outcome == SlotOutcome.ALREADY_RESIDENT
    assert len(builds) == 1


async def test_slot_swap_evicts_incumbent() -> None:
    mgr = ResidentSlotManager(0, lambda: 1_000_000)
    a, b = _FixedGen("A"), _FixedGen("B")
    await mgr.ensure_specialist_async(_sel("A", 1), lambda mid: a)
    await mgr.ensure_specialist_async(_sel("B", 1), lambda mid: b)
    assert a.disposed is True
    assert b.disposed is False
    assert mgr.resident_specialist_model_id == "B"


async def test_slot_build_failure_leaves_slot_empty() -> None:
    mgr = ResidentSlotManager(0, lambda: 1_000_000)

    def boom(mid: str):
        raise RuntimeError("boom")

    a = await mgr.ensure_specialist_async(_sel("spec", 1), boom)
    assert a.outcome == SlotOutcome.BUILD_FAILED
    assert mgr.resident_specialist_model_id is None


async def test_slot_evict_disposes_and_empties() -> None:
    mgr = ResidentSlotManager(0, lambda: 1_000_000)
    g = _FixedGen("S")
    await mgr.ensure_specialist_async(_sel("spec", 1), lambda mid: g)
    await mgr.evict_specialist_async()
    assert g.disposed is True
    assert mgr.resident_specialist_model_id is None


# ── AIService two-slot residency ─────────────────────────────────────────────


async def test_router_none_uses_generalist() -> None:
    svc = AIService(AIOptions(model_id="gen", warm_on_start=False), generator=_FixedGen("GEN"))
    await svc.start_async()
    assert await svc.ask_async("solve this equation") == "GEN"  # reasoning cue, no router


async def test_router_hot_loads_specialist() -> None:
    gen, spec = _FixedGen("GEN"), _FixedGen("SPEC")
    opts = AIOptions(
        model_id="gen",
        warm_on_start=False,
        router=_FixedRouter(RouteDecision.specialist(ChatCapability.REASONING, "t")),
        model_selector=_FixedSelector(_sel("spec-model", 1_024)),
        specialist_factory=lambda mid: spec,
    )
    svc = AIService(opts, generator=gen)
    await svc.start_async()
    assert await svc.ask_async("anything") == "SPEC"


async def test_best_fit_equals_generalist_uses_generalist() -> None:
    gen, spec = _FixedGen("GEN"), _FixedGen("SPEC")
    opts = AIOptions(
        model_id="gen-model",
        warm_on_start=False,
        router=_FixedRouter(RouteDecision.specialist(ChatCapability.REASONING, "t")),
        model_selector=_FixedSelector(_sel("gen-model", 1_024)),
        specialist_factory=lambda mid: spec,
    )
    svc = AIService(opts, generator=gen)
    await svc.start_async()
    assert await svc.ask_async("anything") == "GEN"


async def test_session_round_trip(tmp_path) -> None:
    svc = AIService(AIOptions(model_id="gen", warm_on_start=False), generator=_FixedGen("GEN"))
    await svc.start_async()
    snap = str(tmp_path / "active.session")
    assert await svc.save_session_async(snap) is True
    assert await svc.load_session_async(snap) is True


# ── NeuronNode facade + NullChatRuntime ──────────────────────────────────────


async def test_neuron_node_over_brain(tmp_path) -> None:
    svc = AIService(AIOptions(model_id="qwen-x", warm_on_start=False), generator=_FixedGen("hello"))
    node = NeuronNode(svc)

    assert node.id == "circleai-neuron"
    assert node.is_ready is False
    assert node.status_message == "loading model…"

    await svc.start_async()
    assert node.is_ready is True
    assert node.status_message == "ready"
    assert "qwen-x" in node.engine_label

    chunks = [c async for c in node.stream_async([ChatTurn("user", "hi")])]
    assert "".join(chunks) == "hello"

    snap = str(tmp_path / "active.session")
    assert await node.save_session_async(snap) is True
    assert await node.load_session_async(snap) is True
    assert node.session_snapshot_path


async def test_null_chat_runtime_yields_offline_status() -> None:
    null = NullChatRuntime()
    assert null.is_ready is False
    chunks = [c async for c in null.stream_async([ChatTurn("user", "hi")])]
    assert "No chat engine" in "".join(chunks)

File: c/tests/test_neuron.c
/*
 * test_neuron.c — the Neuron (CircleAI.Hosting.Neuron), C11.
 *
 * Mirrors the C# CircleAI.Tests Neuron suite: the concierge decision table +
 * gate, the two-slot admission gate + eviction, the NeuronNode facade over a
 * brain, and NullChatRuntime.
 */

#include <assert.h>
#include <stdio.h>
#include <string.h>
#include "circle_ai/circle_ai.h"

/* ── test doubles ─────────────────────────────────────────────────────────── */

static int64_t ram_1m(void *u) {
    (void)u;
    return 1000000;
}

static int g_builds = 0;
static int g_token; /* a non-NULL handle */

static void *build_ok(const char *id, void *u) {
    (void)id;
    (void)u;
    g_builds++;
    return &g_token;
}

static void *build_null(const char *id, void *u) {
    (void)id;
    (void)u;
    return NULL;
}

static bool gate_deny(const ca_route_decision_t *d, void *u) {
    (void)d;
    (void)u;
    return false;
}

/* A fake brain for the NeuronNode. */
typedef struct {
    bool        ready;
    const char *model_id;
    const char *reply;
} fake_brain_t;

static bool fb_is_ready(void *impl) { return ((fake_brain_t *)impl)->ready; }
static const char *fb_model_id(void *impl) { return ((fake_brain_t *)impl)->model_id; }
static void fb_stream(void *impl, const ca_chat_turn_t *turns, size_t n,
                      ca_chunk_callback cb, void *ud) {
    (void)turns;
    (void)n;
    if (cb) cb(((fake_brain_t *)impl)->reply, ud);
}
static bool fb_save(void *impl, const char *path) { (void)impl; (void)path; return true; }
static bool fb_load(void *impl, const char *path) { (void)impl; (void)path; return true; }

static char g_out[256];
static void collect(const char *chunk, void *ud) {
    (void)ud;
    strncat(g_out, chunk, sizeof g_out - strlen(g_out) - 1);
}

/* ── tests ────────────────────────────────────────────────────────────────── */

static void test_router(void) {
    ca_neuron_router_t r;
    ca_neuron_router_init(&r);

    /* plain → generalist(DEFAULT) */
    ca_route_context_t plain = { "what's the weather today?", false, -1 };
    ca_route_decision_t d = ca_neuron_route(&r, &plain);
    assert(d.organ == CA_ORGAN_GENERALIST);
    assert(d.capability == CA_CHAT_CAP_DEFAULT);

    /* image → specialist(VISION) */
    ca_route_context_t vis = { "what is this?", true, -1 };
    d = ca_neuron_route(&r, &vis);
    assert(d.organ == CA_ORGAN_SPECIALIST);
    assert(d.capability == CA_CHAT_CAP_VISION);

    /* reasoning cue → specialist(REASONING) */
    ca_route_context_t rea = { "please debug this stack trace", false, -1 };
    d = ca_neuron_route(&r, &rea);
    assert(d.organ == CA_ORGAN_SPECIALIST);
    assert(d.capability == CA_CHAT_CAP_REASONING);

    /* long prompt → specialist(LONG_CTX) */
    ca_neuron_router_t rl;
    ca_neuron_router_init(&rl);
    rl.long_context_chars = 50;
    char big[61];
    memset(big, 'x', 60);
    big[60] = '\0';
    ca_route_context_t lng = { big, false, -1 };
    d = ca_neuron_route(&rl, &lng);
    assert(d.organ == CA_ORGAN_SPECIALIST);
    assert(d.capability == CA_CHAT_CAP_LONG_CTX);

    /* gate veto demotes a specialist back to the generalist */
    ca_neuron_router_t rg;
    ca_neuron_router_init(&rg);
    rg.gate = gate_deny;
    ca_route_context_t solve = { "solve this equation", false, -1 };
    d = ca_neuron_route(&rg, &solve);
    assert(d.organ == CA_ORGAN_GENERALIST);
}

static void test_slot_manager(void) {
    ca_resident_slot_manager_t m;
    void *out = NULL;

    /* admits within budget */
    ca_slot_manager_init(&m, 1000, ram_1m, NULL);
    assert(ca_slot_ensure_specialist(&m, "spec", 5000, build_ok, NULL, &out) == CA_SLOT_ADMITTED);
    assert(out == &g_token);
    assert(ca_slot_resident_model_id(&m) && strcmp(ca_slot_resident_model_id(&m), "spec") == 0);

    /* denies over budget */
    ca_slot_manager_init(&m, 900000, ram_1m, NULL);
    assert(ca_slot_ensure_specialist(&m, "spec", 500000, build_ok, NULL, &out) == CA_SLOT_INSUFFICIENT_RAM);
    assert(ca_slot_resident_model_id(&m) == NULL);

    /* already resident → build once */
    ca_slot_manager_init(&m, 0, ram_1m, NULL);
    g_builds = 0;
    assert(ca_slot_ensure_specialist(&m, "spec", 1, build_ok, NULL, NULL) == CA_SLOT_ADMITTED);
    assert(ca_slot_ensure_specialist(&m, "spec", 1, build_ok, NULL, NULL) == CA_SLOT_ALREADY_RESIDENT);
    assert(g_builds == 1);

    /* a different pick evicts the incumbent */
    ca_slot_manager_init(&m, 0, ram_1m, NULL);
    ca_slot_ensure_specialist(&m, "A", 1, build_ok, NULL, NULL);
    ca_slot_ensure_specialist(&m, "B", 1, build_ok, NULL, NULL);
    assert(strcmp(ca_slot_resident_model_id(&m), "B") == 0);

    /* build failure leaves the slot empty */
    ca_slot_manager_init(&m, 0, ram_1m, NULL);
    assert(ca_slot_ensure_specialist(&m, "spec", 1, build_null, NULL, NULL) == CA_SLOT_BUILD_FAILED);
    assert(ca_slot_resident_model_id(&m) == NULL);

    /* evict clears the slot */
    ca_slot_manager_init(&m, 0, ram_1m, NULL);
    ca_slot_ensure_specialist(&m, "spec", 1, build_ok, NULL, NULL);
    ca_slot_evict(&m);
    assert(ca_slot_resident_model_id(&m) == NULL);
}

static void test_neuron_node(void) {
    fake_brain_t fb = { false, "qwen-x", "hello" };
    ca_neuron_brain_t brain = {
        fb_is_ready, fb_model_id, fb_stream, fb_save, fb_load, &fb
    };
    ca_neuron_node_t node;
    ca_neuron_node_init(&node, brain, NULL);

    assert(strcmp(ca_neuron_node_id(&node), "circleai-neuron") == 0);
    assert(!ca_neuron_node_is_ready(&node));
    assert(strstr(ca_neuron_node_status(&node), "loading") != NULL);

    fb.ready = true;
    assert(ca_neuron_node_is_ready(&node));
    assert(strcmp(ca_neuron_node_status(&node), "ready") == 0);
    assert(strstr(ca_neuron_node_engine_label(&node), "qwen-x") != NULL);

    g_out[0] = '\0';
    ca_chat_turn_t turns[1] = { { "user", "hi" } };
    ca_neuron_node_stream(&node, turns, 1, collect, NULL);
    assert(strcmp(g_out, "hello") == 0);

    assert(ca_neuron_node_snapshot_path(&node) != NULL);
    assert(ca_neuron_node_save_session(&node, "/tmp/x") == true);
    assert(ca_neuron_node_load_session(&node, "/tmp/x") == true);
}

static void test_null_runtime(void) {
    assert(ca_null_chat_runtime_is_ready() == false);
    g_out[0] = '\0';
    ca_chat_turn_t turns[1] = { { "user", "hi" } };
    ca_null_chat_runtime_stream(turns, 1, collect, NULL);
    assert(strstr(g_out, "No chat engine") != NULL);
}

int main(void) {
    test_router();
    test_slot_manager();
    test_neuron_node();
    test_null_runtime();
    printf("test_neuron: all assertions passed\n");
    return 0;
}
